### PR TITLE
Feat/225 create model for user

### PR DESCRIPTION
**Pull request by matvitn**
# Implement User Model and Firestore Integration

## Description

This PR implements the User model and integrates it with Firebase Firestore, addressing issue [[#225](https://github.com/commit-crimes/shelf-life/issues/225)](https://github.com/commit-crimes/shelf-life/issues/225). The key changes include:

- **Added `User` data class**: Represents the user entity with fields for `uid`, `username`, `email`, `householdUIDs`, `recipeUIDs`, and `invitationUIDs`.

- **Implemented `UserRepository` interface**: Defines the contract for user data operations.

- **Created `UserRepositoryFirestore` class**: Implements the `UserRepository` interface using Firebase Firestore for data persistence.

- **Synchronized field names**: Updated code to ensure consistency between Firestore document fields and the `User` data class properties.

- **Wrote unit tests for `UserRepositoryFirestore`**: Achieved close to 80% code coverage, testing various methods and scenarios.

- **Refactored code for maintainability**: Improved code structure and formatting for better readability and consistency.

### New Files Added

- `app/src/main/java/com/android/shelfLife/model/user/User.kt`
  - Defines the `User` data class with necessary properties and a custom `toString` method.

- `app/src/main/java/com/android/shelfLife/model/user/UserRepository.kt`
  - Declares the `UserRepository` interface with methods for user data operations.

- `app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt`
  - Implements the `UserRepository` interface using Firebase Firestore.

- `app/src/test/java/com/android/shelflife/model/user/UserRepositoryFirestoreTest.kt`
  - Contains unit tests for the `UserRepositoryFirestore` class.

### Modifications

- **Formatted `EditFoodItem.kt` and `FoodItemFields.kt`**:
  - Applied consistent code formatting for better readability.

- **Updated field names in code**:
  - Ensured all field names match those in the `User` data class (`householdUIDs`, `recipeUIDs`, `invitationUIDs`).

## Testing

- **Unit Tests**:
  - Wrote comprehensive unit tests for `UserRepositoryFirestore`, covering methods like `initializeUserData`, `addHouseholdUID`, `deleteHouseholdUID`, `updateUsername`, and `updateEmail`.
  - Used Mockito for mocking Firebase dependencies.
  - Employed reflection in tests to inject mocked instances into private fields without altering production code.

- **Test Coverage**:
  - Achieved close to 80% code coverage on the `UserRepositoryFirestore` class.

## Notes for Reviewers

- **Firebase Mocking**:
  - FirebaseAuth and Firestore are mocked using Mockito.
  - Private fields are accessed via reflection in tests to inject mocks, avoiding changes to the production code.

- **Considerations for Future Improvements**:
  - Exploring dependency injection frameworks could enhance testability and code maintainability.
  - Using a Kotlin-friendly mocking library like MockK may simplify testing coroutine-based code.

## Related Issues

- Closes [[#225](https://github.com/commit-crimes/shelf-life/issues/225)](https://github.com/commit-crimes/shelf-life/issues/225)

---

Please review the implementation and let me know if any adjustments are needed.

### CODE REVIEW

File: app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        when (fieldName) {
+          "username" -> currentUserData.copy(username = value as String)
+          "email" -> currentUserData.copy(email = value as String)
+          else -> currentUserData
+        }
+    _user.value = updatedUserData
+  }
+
+  // Helper function to update array fields and local user data
+  private suspend fun updateArrayField(
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt:105]: good job modularising this, and keeping Recipe, Household CRUD functions clean

File: app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  override suspend fun updateUsername(username: String) {
+    updateUserField("username", username)
+  }
+
+  override suspend fun updateEmail(email: String) {
+    val currentUser = auth.currentUser ?: throw Exception("User not logged in")
+    // Update email in FirebaseAuth
+    currentUser.updateEmail(email).await()
+    // Update email in Firestore
+    updateUserField("email", email)
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt:183]: could put this in a try catch in case the firebase update fails (connection error, etc..)
```suggestion
    try{
      currentUser.updateEmail(email).await()
      // Update email in Firestore
      updateUserField("email", email)
    }catch(e: FirebaseAuthException) {
        Log.e("UserRepository", "Failed to update email in FirebaseAuth", e)
        throw e
    }
```

File: app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+  override suspend fun deleteRecipeUID(uid: String) {
+    updateArrayField("recipeUIDs", uid, ArrayOperation.REMOVE)
+  }
+
+  override suspend fun deleteInvitationUID(uid: String) {
+    // Update Firestore
+    updateArrayField("invitationUIDs", uid, ArrayOperation.REMOVE)
+    // No need to update local _user or _invitations, as listener will handle it
+  }
+
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt:173]: ## Tests
Missing tests for RecipeUIDs functions add and delete. You can make some new tests that are rather similar to the Household ones you already made

File: app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+          }
+    } else {
+      _invitations.value = emptyList()
+    }
+  }
+
+  override fun stopListeningForInvitations() {
+    invitationsListenerRegistration?.remove()
+    invitationsListenerRegistration = null
+  }
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt:86]: ## Tests
maybe add a test for this function. Something along the lines of 

```kotlin
@Test
fun `stopListeningForInvitations stops the listener correctly`() {
    val mockListenerRegistration: ListenerRegistration = mock()

    `when`(mockUserCollection.addSnapshotListener(any())).thenReturn(mockListenerRegistration)

    userRepository.stopListeningForInvitations()

    verify(mockListenerRegistration).remove()
}

```

File: app/src/test/java/com/android/shelflife/model/user/UserRepositoryFirestoreTest.kt [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+  @Test
+  fun `addHouseholdUID adds household UID to Firestore and updates local user`() = runTest {
+    val initialUser =
+        User(
+            uid = "testUserId",
+            username = "TestUser",
+            email = "test@example.com",
+            householdUIDs = mutableListOf("house1"))
+    // Access the private _user variable via reflection or adjust visibility
+    val userField = userRepository.javaClass.getDeclaredField("_user")
```
Comment (alextheprophecy) [inline app/src/test/java/com/android/shelflife/model/user/UserRepositoryFirestoreTest.kt:133]: Good job using reflection to access the private field!!

**Review by alextheprophecy — APPROVED** [review]

File: app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+      _invitations.value = emptyList()
+      throw e
+    }
+  }
+
+  override fun startListeningForInvitations() {
+    val currentUser = auth.currentUser
+    if (currentUser != null) {
+      // Remove any existing listener
+      invitationsListenerRegistration?.remove()
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/model/user/UserRepositoryFirestore.kt:58]: nit picky: maybe it's better to call `stopListeningForInvitations()` here? Maybe there is a reason you explicitly wrote the removal line again

**Comment by alextheprophecy** [comment]
## Review Summary

Overall, this implementation is well done and aligns with what we discussed in our scrum meetings regarding the **big Model overhaul**. Below are some key points:

- **Well-structured code**: The repository pattern is well implemented with clear focus on Firebase integration.
- **Functionality**: Core functionality is clear, with the correct listener implementations and updating the Repository local list

### Areas for Improvement

- **Missing tests**: A few tests are still missing, particularly for methods like:
  - `addRecipeUID`
  - `deleteRecipeUID`
  - removing the listener
- **Test coverage**: Adding these missing tests would improve overall test coverage, especially for edge cases.

### Good Job- Template for Future Models

- This model class can serve as a **template** for future models in the new refactoring effort.
- **It sets a solid foundation for working with state flows in the**Firestore** repositories and keeping data up to date with listeners**
